# Table of Contents
 <p><div class="lev1 toc-item"><a href="#MIDS---w261-Machine-Learning-At-Scale" data-toc-modified-id="MIDS---w261-Machine-Learning-At-Scale-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>MIDS - w261 Machine Learning At Scale</a></div><div class="lev2 toc-item"><a href="#Assignment---HW5" data-toc-modified-id="Assignment---HW5-11"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Assignment - HW5</a></div><div class="lev1 toc-item"><a href="#Table-of-Contents-" data-toc-modified-id="Table-of-Contents--2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Table of Contents </a></div><div class="lev1 toc-item"><a href="#1-Instructions" data-toc-modified-id="1-Instructions-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>1 Instructions</a></div><div class="lev3 toc-item"><a href="#IMPORTANT" data-toc-modified-id="IMPORTANT-301"><span class="toc-item-num">3.0.1&nbsp;&nbsp;</span>IMPORTANT</a></div><div class="lev3 toc-item"><a href="#===-INSTRUCTIONS-for-SUBMISSIONS-===" data-toc-modified-id="===-INSTRUCTIONS-for-SUBMISSIONS-===-302"><span class="toc-item-num">3.0.2&nbsp;&nbsp;</span>=== INSTRUCTIONS for SUBMISSIONS ===</a></div><div class="lev1 toc-item"><a href="#2-Useful-References" data-toc-modified-id="2-Useful-References-4"><span class="toc-item-num">4&nbsp;&nbsp;</span></a></div><div class="lev1 toc-item"><a href="#HW-Problems" data-toc-modified-id="HW-Problems-5"><span class="toc-item-num">5&nbsp;&nbsp;</span></a></div><div class="lev2 toc-item"><a href="#3.--HW5.0--data-warehouse;-star-schema" data-toc-modified-id="3.--HW5.0--data-warehouse;-star-schema-51"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>3.  HW5.0  data warehouse; star schema</a></div><div class="lev2 toc-item"><a href="#3.--HW5.1-Databases:-3NF;-denormalized-" data-toc-modified-id="3.--HW5.1-Databases:-3NF;-denormalized--52"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>3.  HW5.1 Databases: 3NF; denormalized </a></div><div class="lev2 toc-item"><a href="#3.--HW5.2--Memory-backed-map-side" data-toc-modified-id="3.--HW5.2--Memory-backed-map-side-53"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>3.  HW5.2  Memory-backed map-side</a></div><div class="lev2 toc-item"><a href="#3.--HW5.2.1-(OPTIONAL)-Almost-stateless-reducer-side-join--" data-toc-modified-id="3.--HW5.2.1-(OPTIONAL)-Almost-stateless-reducer-side-join---54"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>3.  HW5.2.1 (OPTIONAL) Almost stateless reducer-side join  </a></div><div class="lev1 toc-item"><a href="#5.3-Pairwise-similarity----PHASE-1-" data-toc-modified-id="5.3-Pairwise-similarity----PHASE-1--6"><span class="toc-item-num">6&nbsp;&nbsp;</span>5.3 Pairwise similarity  - PHASE 1 </a></div><div class="lev4 toc-item"><a href="#(1)-Using-the-systems-tests-data-sets,-write-mrjob-code-to-build-the-stripes" data-toc-modified-id="(1)-Using-the-systems-tests-data-sets,-write-mrjob-code-to-build-the-stripes-6001"><span class="toc-item-num">6.0.0.1&nbsp;&nbsp;</span>(1) Using the systems tests data sets, write mrjob code to build the stripes</a></div><div class="lev4 toc-item"><a href="#(2)-Write-mrjob-code-to-build-an-inverted-index-from-the-stripes" data-toc-modified-id="(2)-Write-mrjob-code-to-build-an-inverted-index-from-the-stripes-6002"><span class="toc-item-num">6.0.0.2&nbsp;&nbsp;</span>(2) Write mrjob code to build an inverted index from the stripes</a></div><div class="lev4 toc-item"><a href="#(3)-Using-two-(symmetric)-comparison-methods-of-your-choice-(e.g.,-correlations,-distances,-similarities),-pairwise-compare-all-stripes-(vectors),-and-output-to-a-file." data-toc-modified-id="(3)-Using-two-(symmetric)-comparison-methods-of-your-choice-(e.g.,-correlations,-distances,-similarities),-pairwise-compare-all-stripes-(vectors),-and-output-to-a-file.-6003"><span class="toc-item-num">6.0.0.3&nbsp;&nbsp;</span>(3) Using two (symmetric) comparison methods of your choice (e.g., correlations, distances, similarities), pairwise compare all stripes (vectors), and output to a file.</a></div><div class="lev2 toc-item"><a href="#HW5.3.1---Run-Systems-tests-locally-on-small-datasets-(PHASE1)-" data-toc-modified-id="HW5.3.1---Run-Systems-tests-locally-on-small-datasets-(PHASE1)--61"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>HW5.3.1   Run Systems tests locally on small datasets (PHASE1) </a></div><div class="lev4 toc-item"><a href="#1:-unit/systems-first-10-lines" data-toc-modified-id="1:-unit/systems-first-10-lines-6101"><span class="toc-item-num">6.1.0.1&nbsp;&nbsp;</span>1: unit/systems first-10-lines</a></div><div class="lev4 toc-item"><a href="#2:-unit/systems-atlas-boon" data-toc-modified-id="2:-unit/systems-atlas-boon-6102"><span class="toc-item-num">6.1.0.2&nbsp;&nbsp;</span>2: unit/systems atlas-boon</a></div><div class="lev4 toc-item"><a href="#3:-unit/systems-stripe-docs-test" data-toc-modified-id="3:-unit/systems-stripe-docs-test-6103"><span class="toc-item-num">6.1.0.3&nbsp;&nbsp;</span>3: unit/systems stripe-docs-test</a></div><div class="lev3 toc-item"><a href="#(1)-build-stripes-for-all-the-test-data-sets---run-the-commands-and-insure-that-your-output-matches-the-output-below" data-toc-modified-id="(1)-build-stripes-for-all-the-test-data-sets---run-the-commands-and-insure-that-your-output-matches-the-output-below-611"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>(1) build stripes for all the test data sets - run the commands and insure that your output matches the output below</a></div><div class="lev3 toc-item"><a href="#(2)-Build-Inverted-Index---run-the-commands-and-insure-that-your-output-matches-the-output-below" data-toc-modified-id="(2)-Build-Inverted-Index---run-the-commands-and-insure-that-your-output-matches-the-output-below-612"><span class="toc-item-num">6.1.2&nbsp;&nbsp;</span>(2) Build Inverted Index - run the commands and insure that your output matches the output below</a></div><div class="lev3 toc-item"><a href="#Inverted-Index" data-toc-modified-id="Inverted-Index-613"><span class="toc-item-num">6.1.3&nbsp;&nbsp;</span>Inverted Index</a></div><div class="lev3 toc-item"><a href="#(3)-Calculate-similarities---run-the-commands-and-insure-that-your-output-matches-the-output-below" data-toc-modified-id="(3)-Calculate-similarities---run-the-commands-and-insure-that-your-output-matches-the-output-below-614"><span class="toc-item-num">6.1.4&nbsp;&nbsp;</span>(3) Calculate similarities - run the commands and insure that your output matches the output below</a></div><div class="lev4 toc-item"><a href="#NOTE:-you-must-run-in-hadoop-mode-to-generate-sorted-similarities" data-toc-modified-id="NOTE:-you-must-run-in-hadoop-mode-to-generate-sorted-similarities-6141"><span class="toc-item-num">6.1.4.1&nbsp;&nbsp;</span>NOTE: you must run in hadoop mode to generate sorted similarities</a></div><div class="lev3 toc-item"><a href="#Pairwise-Similairity" data-toc-modified-id="Pairwise-Similairity-615"><span class="toc-item-num">6.1.5&nbsp;&nbsp;</span>Pairwise Similairity</a></div><div class="lev1 toc-item"><a href="#===-END-OF-PHASE-1-===" data-toc-modified-id="===-END-OF-PHASE-1-===-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>=== END OF PHASE 1 ===</a></div>

# MIDS - w261 Machine Learning At Scale
__Course Lead:__ Dr James G. Shanahan (__email__ Jimi via  James.Shanahan _AT_ gmail.com)

## Assignment - HW5


---
__Name:__  Nicholas W. Chen   
__Class:__ MIDS w261 (Section Summer 2017 Group 1)     
__Email:__  nwchen24@iSchool.Berkeley.edu     
__StudentId__  3032079012    __End of StudentId__     
__Week:__   5

__NOTE:__ please replace `1234567` with your student id above      
__Due Time:__ HW is due the Tuesday of the following week by 8AM (West coast time). I.e., Tuesday, Feb 14, 2017 in the case of this homework. 

* __HW5 Phase 1__ 
This can be done on a local machine (with a unit test on the cloud such as AltaScale's PaaS or on AWS) and is due Tuesday, Week 6 by 8AM (West coast time). It will primarily focus on building a unit/systems and for pairwise similarity calculations pipeline (for stripe documents)

* __HW5 Phase 2__ 
This will require the AltaScale cluster and will be due Tuesday, Feb 21 by 8AM (West coast time). 
The focus of  HW5 Phase 2  will be to scale up the unit/systems tests to the Google 5 gram corpus. This will be a group exercise 


# Table of Contents <a name="TOC"></a> 

1.  [HW Intructions](#1)   
2.  [HW References](#2)
3.  [HW Problems](#3)   
1.  [HW Introduction](#1)   
2.  [HW References](#2)
3.  [HW  Problems](#3)   
    5.0.  [HW5.0](#5.0)   
    5.1.  [HW5.1](#5.1)   
    5.2.  [HW5.2](#5.2)   
    5.3.  [HW5.3](#5.3)    
    
   

<a name="1"></a>
# 1 Instructions
[Back to Table of Contents](#TOC)

MIDS UC Berkeley, Machine Learning at Scale   
DATSCIW261 ASSIGNMENT #5

Version 2017-9-2 


### IMPORTANT

This homework can be completed locally on your computer 

### === INSTRUCTIONS for SUBMISSIONS ===   
Follow the instructions for submissions carefully.

Each student has a `HW-<user>` repository for all assignments.   

Click this link to enable you to create a github repo within the MIDS261 Classroom:   
https://classroom.github.com/assignment-invitations/3b1d6c8e58351209f9dd865537111ff8   
and follow the instructions to create a HW repo.

Push the following to your HW github repo into the master branch:
* Your local HW5 directory. Your repo file structure should look like this:

```
HW-<user>
    --HW3
       |__MIDS-W261-HW-03-<Student_id>.ipynb
       |__MIDS-W261-HW-03-<Student_id>.pdf
       |__some other hw3 file
    --HW4
       |__MIDS-W261-HW-04-<Student_id>.ipynb
       |__MIDS-W261-HW-04-<Student_id>.pdf
       |__some other hw4 file
    etc..
```    

  
<a name="2">
# 2 Useful References
[Back to Table of Contents](#TOC)

* See async and live lectures for this week

<a name="3">
# HW Problems
[Back to Table of Contents](#TOC)

## 3.  HW5.0  data warehouse; star schema<a name="5.0"></a>
[Back to Table of Contents](#TOC)

- What is a data warehouse? What is a Star schema? When is it used?

Data warehouses are where organizations store their data. Historically, data warehouses were typically relational in nature and often organized using a Star schema. Today, however, data warehouses are evolving and are being used to store both structured and unstructured data.  

A Star schema is a way of organizing a relational database when the database is organized around a central table that contains various dimensions about a given metric.

## 3.  HW5.1 Databases: 3NF; denormalized <a name="5.1"></a>
[Back to Table of Contents](#TOC)

- In the database world What is 3NF? Does machine learning use data in 3NF? If so why? 
- In what form does ML consume data?
- Why would one use log files that are denormalized?

3NF stands for third normal form. This means that the data are stored in separate tables that relate to one another by various identifiers. The reason for storing data in third normal form is to minimize data duplication.  

Machine learning does not use data in third normal form because it needs to have all of the information for each record available at the same time. In third normal form, the information for a single record can be distributed across numerous tables.  

ML consumes data in denormalized form. This means that the numerous tables in third normal form have been recombined into a single record so that each record fed to the ML algorithm is complete.  

One would use log files that are denormalized so that each record is complete and can be processed by an ML algorithm. A downside of denormalized form is that it can take up much more space than data stored in third normal form because  data is duplicated in denormalized form.

## 3.  HW5.2  Memory-backed map-side<a name="5.2"></a>
[Back to Table of Contents](#TOC)

Using MRJob, implement a hashside join (memory-backed map-side) for left, right and inner joins. Use the following tables for this HW and join based on the country code (third column of the transactions table and the second column of the Countries table:

<PRE>
transactions.dat
Alice Bob|$10|US
Sam Sneed|$1|CA
Jon Sneed|$20|CA
Arnold Wesise|$400|UK
Henry Bob|$2|US
Yo Yo Ma|$2|CA
Jon York|$44|CA
Alex Ball|$5|UK
Jim Davis|$66|JA

Countries.dat
United States|US
Canada|CA
United Kingdom|UK
Italy|IT

</PRE>

Justify which table you chose as the Left table in this hashside join.

Please report the number of rows resulting from:

- (1) Left joining Table Left with Table Right
- (2) Right joining Table Left with Table Right
- (3) Inner joining Table Left with Table Right


### Write Tables

In [10]:
%%writefile transactions.dat
Alice Bob|$10|US
Sam Sneed|$1|CA
Jon Sneed|$20|CA
Arnold Wesise|$400|UK
Henry Bob|$2|US
Yo Yo Ma|$2|CA
Jon York|$44|CA
Alex Ball|$5|UK
Jim Davis|$66|JA



Overwriting transactions.dat


In [3]:
%%writefile Countries.dat
United States|US
Canada|CA
United Kingdom|UK
Italy|IT

Writing Countries.dat


### Left Side Join

In [114]:
%%writefile Left_join_52.py
#!/opt/anaconda/bin/python

from mrjob.job import MRJob
from mrjob.step import MRStep
import ast

class MR_Left_Join_52(MRJob):
    
    # OUTPUT_PROTOCOL = JSONValueProtocol
    MRJob.SORT_VALUES = True

    #For in-memory mapper-side join, read Countries data into memory
    def mapper_init(self):

        self.countries_file = open("Countries.dat").readlines()
        self.countries_dict = {}

        for line in self.countries_file:
            #parse each line
            Country_full,Country_abbrev = line.split("|")
            #update the dict with the abbreviated verision as the key
            self.countries_dict[Country_abbrev.strip().translate(None,"\n")] = Country_full

    def mapper(self, _, line):
        #read the input
        Customer,Amount,Country_abbrev = line.split("|")
        
        #This is the join step
        try:
            #get the full country from the dict
            Country_full = self.countries_dict[Country_abbrev.strip().translate(None,"\n")]
            
            #Yield the joined records
            yield Customer, (Country_full, Amount, Country_abbrev)
        
        #don't emit anything if the country abbreviation not found in the dict - this makes this join an inner join
        except KeyError:
            pass
        
        
    def reducer_init_written(self):

        #Create a dict where we will keep track of which countries have not been encountered in the transaction data
        self.countries_encountered_dict = {}

        self.countries_file = open("Countries.dat").readlines()

        for line in self.countries_file:
            #parse each line
            Country_full,Country_abbrev = line.split("|")
            #update the dict with the abbreviated verision as the key
            self.countries_encountered_dict[Country_abbrev.strip().translate(None,"\n")] = Country_full
        #yield customer, input_data
                  
    def reducer(self, customer, inputdata):
        
        #When a country code is encountered, remove it from the countries file
        for Country_full, Amount, Country_abbrev in inputdata:
        
            try:
                del self.countries_encountered_dict[Country_abbrev]
            except KeyError:
                pass
        
        #emit the record passed from the mapper
        yield customer, (Country_full, Amount, Country_abbrev)

    def reducer_final(self):
        
        #if there are still country entries in the dict that have not been encountered, emit these (this step completes the left join)
        for key in self.countries_encountered_dict:
            yield None, (self.countries_encountered_dict[key],None,key)
       
    def steps(self):

        return [
            MRStep(mapper_init=self.mapper_init,
                   mapper=self.mapper,
                   reducer_init=self.reducer_init_written,
                   reducer=self.reducer,
                   reducer_final=self.reducer_final
                  )
                ]
          
if __name__ == '__main__':
    MR_Left_Join_52.run()

Overwriting Left_join_52.py


In [115]:
#Make the file executable
!chmod +x Left_join_52.py

In [116]:
#Run the job
!python Left_join_52.py -r hadoop --file=Countries.dat transactions.dat > Left_join_52_Output.txt

Using configs in /etc/mrjob.conf
Looking for hadoop binary in $PATH...
Found hadoop binary: /usr/bin/hadoop
Using Hadoop version 2.6.0
Looking for Hadoop streaming jar in /home/hadoop/contrib...
Looking for Hadoop streaming jar in /usr/lib/hadoop-mapreduce...
Found Hadoop streaming jar: /usr/lib/hadoop-mapreduce/hadoop-streaming.jar
Creating temp directory /tmp/Left_join_52.root.20170608.143129.465678
Copying local files to hdfs:///user/root/tmp/mrjob/Left_join_52.root.20170608.143129.465678/files/...
Detected hadoop configuration property names that do not match hadoop version 2.6.0:
The have been translated as follows
 mapred.output.key.comparator.class: mapreduce.job.output.key.comparator.class
mapred.text.key.comparator.options: mapreduce.partition.keycomparator.options
mapred.text.key.partitioner.options: mapreduce.partition.keypartitioner.options
Running step 1 of 1...
  mapred.text.key.partitioner.options is deprecated. Instead, use mapreduce.partition.keypartitioner.options
  p

In [117]:
!head Left_join_52_Output.txt

"Alex Ball"	["United Kingdom","$5","UK"]
"Alice Bob"	["United States","$10","US"]
"Arnold Wesise"	["United Kingdom","$400","UK"]
"Henry Bob"	["United States","$2","US"]
"Jon Sneed"	["Canada","$20","CA"]
"Jon York"	["Canada","$44","CA"]
"Sam Sneed"	["Canada","$1","CA"]
"Yo Yo Ma"	["Canada","$2","CA"]
null	["Italy",null,"IT"]


### Inner Join

In [5]:
%%writefile Inner_join_52.py
#!/opt/anaconda/bin/python

from mrjob.job import MRJob
from mrjob.step import MRStep

class MR_Inner_Join_52(MRJob):
    
    MRJob.SORT_VALUES = True

    #For in-memory mapper-side join, read Countries data into memory
    def mapper_init(self):

        self.countries_file = open("Countries.dat").readlines()
        self.countries_dict = {}

        for line in self.countries_file:
            #parse each line
            Country_full,Country_abbrev = line.split("|")
            #update the dict with the abbreviated verision as the key
            self.countries_dict[Country_abbrev.strip().translate(None,"\n")] = Country_full

    def mapper(self, _, line):
        #read the input
        Customer,Amount,Country_abbrev = line.split("|")
        
        #This is the join step
        try:
            #get the full country from the dict
            Country_full = self.countries_dict[Country_abbrev.strip().translate(None,"\n")]
            #Yield the joined records
            yield Customer, (Country_full, Amount, Country_abbrev)
        
        #don't emit anything if the country abbreviation not found in the dict - this makes this join an inner join
        except KeyError:
            pass

    def steps(self):

        return [
            MRStep(mapper_init=self.mapper_init,
                   mapper=self.mapper
                  )
                ]
          
if __name__ == '__main__':
    MR_Inner_Join_52.run()

Overwriting Inner_join_52.py


In [6]:
#Make the file executable
!chmod +x Inner_join_52.py

In [7]:
#Run the job
!python Inner_join_52.py -r hadoop --file=Countries.dat transactions.dat > Inner_Join_52_Output.txt

Using configs in /etc/mrjob.conf
Looking for hadoop binary in $PATH...
Found hadoop binary: /usr/bin/hadoop
Using Hadoop version 2.6.0
Looking for Hadoop streaming jar in /home/hadoop/contrib...
Looking for Hadoop streaming jar in /usr/lib/hadoop-mapreduce...
Found Hadoop streaming jar: /usr/lib/hadoop-mapreduce/hadoop-streaming.jar
Creating temp directory /tmp/Inner_join_52.root.20170608.120132.316834
Copying local files to hdfs:///user/root/tmp/mrjob/Inner_join_52.root.20170608.120132.316834/files/...
Detected hadoop configuration property names that do not match hadoop version 2.6.0:
The have been translated as follows
 mapred.output.key.comparator.class: mapreduce.job.output.key.comparator.class
mapred.text.key.comparator.options: mapreduce.partition.keycomparator.options
mapred.text.key.partitioner.options: mapreduce.partition.keypartitioner.options
Running step 1 of 1...
  mapred.text.key.partitioner.options is deprecated. Instead, use mapreduce.partition.keypartitioner.options
 

In [8]:
!head Inner_Join_52_Output.txt

"Alice Bob"	["United States","$10","US"]	
"Sam Sneed"	["Canada","$1","CA"]	
"Jon Sneed"	["Canada","$20","CA"]	
"Arnold Wesise"	["United Kingdom","$400","UK"]	
"Henry Bob"	["United States","$2","US"]	
"Yo Yo Ma"	["Canada","$2","CA"]	
"Jon York"	["Canada","$44","CA"]	
"Alex Ball"	["United Kingdom","$5","UK"]	


### Right Join

In [13]:
%%writefile Right_join_52.py
#!/opt/anaconda/bin/python

from mrjob.job import MRJob
from mrjob.step import MRStep

class MR_Right_Join_52(MRJob):
    
    MRJob.SORT_VALUES = True

    #For in-memory mapper-side join, read Countries data into memory
    def mapper_init(self):

        self.countries_file = open("Countries.dat").readlines()
        self.countries_dict = {}

        for line in self.countries_file:
            #parse each line
            Country_full,Country_abbrev = line.split("|")
            #update the dict with the abbreviated verision as the key
            self.countries_dict[Country_abbrev.strip().translate(None,"\n")] = Country_full

    def mapper(self, _, line):
        #read the input
        Customer,Amount,Country_abbrev = line.split("|")
        
        #This is the join step
        try:
            #get the full country from the dict
            Country_full = self.countries_dict[Country_abbrev.strip().translate(None,"\n")]
            #Yield the joined records
            yield Customer, (Country_full, Amount, Country_abbrev)
        
        #emit the transaction line even if the country abbreviation not found in the dict - this makes this join a right join
        except KeyError:
            yield Customer, (None, Amount, Country_abbrev.strip().translate(None,"\n"))

    def steps(self):

        return [
            MRStep(mapper_init=self.mapper_init,
                   mapper=self.mapper
                  )
                ]
          
if __name__ == '__main__':
    MR_Right_Join_52.run()

Overwriting Right_join_52.py


In [14]:
#Make the file executable
!chmod +x Right_join_52.py

In [15]:
#Run the job
!python Right_join_52.py -r hadoop --file=Countries.dat transactions.dat > Right_join_52_Output.txt

Using configs in /etc/mrjob.conf
Looking for hadoop binary in $PATH...
Found hadoop binary: /usr/bin/hadoop
Using Hadoop version 2.6.0
Looking for Hadoop streaming jar in /home/hadoop/contrib...
Looking for Hadoop streaming jar in /usr/lib/hadoop-mapreduce...
Found Hadoop streaming jar: /usr/lib/hadoop-mapreduce/hadoop-streaming.jar
Creating temp directory /tmp/Right_join_52.root.20170608.120833.725291
Copying local files to hdfs:///user/root/tmp/mrjob/Right_join_52.root.20170608.120833.725291/files/...
Detected hadoop configuration property names that do not match hadoop version 2.6.0:
The have been translated as follows
 mapred.output.key.comparator.class: mapreduce.job.output.key.comparator.class
mapred.text.key.comparator.options: mapreduce.partition.keycomparator.options
mapred.text.key.partitioner.options: mapreduce.partition.keypartitioner.options
Running step 1 of 1...
  mapred.text.key.partitioner.options is deprecated. Instead, use mapreduce.partition.keypartitioner.options
 

In [16]:
!head Right_join_52_Output.txt

"Alice Bob"	["United States","$10","US"]	
"Sam Sneed"	["Canada","$1","CA"]	
"Jon Sneed"	["Canada","$20","CA"]	
"Arnold Wesise"	["United Kingdom","$400","UK"]	
"Henry Bob"	["United States","$2","US"]	
"Yo Yo Ma"	["Canada","$2","CA"]	
"Jon York"	["Canada","$44","CA"]	
"Alex Ball"	["United Kingdom","$5","UK"]	
"Jim Davis"	[null,"$66","JA"]	


## 3.  HW5.2.1 (OPTIONAL) Almost stateless reducer-side join  <a name="5.2.1"></a>
[Back to Table of Contents](#TOC)

The following MRJob code, implements a reduce-side join for an inner join. The reducer is almost stateless, i.e., uses as little memory as possible. Use the tables from HW5.2 for this HW and join based on the country code (third column of the transactions table and the second column of the Countries table perform. Perform  an left, right, inner joins using the code provided below and report the number of rows resulting from:

- (1) Left joining Table Left with Table Right
- (2) Right joining Table Left with Table Right
- (3) Inner joining Table Left with Table Right

Again make smart decisions about which table should be the left table (i.e., crosscheck the code). 

__Some notes on the code__ 
Here, the mapper receives its set of input splits either from the transaction table or from the countries table and makes the appropriate transformations: splitting the line into fields, and emitting a key/value. The key is the join key - in this case, the country code field of both sets of records. The mapper knows which file and type of record it is receiving based on the length of the fields. The records it emits contain the join field as the key, which acts as the partitioning key; We use the SORT_VALUES option, which ensures the values are sorted as well. Then, we employ a trick to ensure that for each join key, country records are seen always before transaction records. We achieve this by adding an arbitrary key to the front of the value: 'A' for countries, 'B' for customers. This makes countries sort before customers for each and every join/partition key. After that trick, the join is simply a matter of storing countries ('A' records) and crossing this array with each customer record.

In [ ]:
import sys, os, re
from mrjob.job import MRJob

class MRJoin(MRJob):

  # Performs secondary sort
  SORT_VALUES = True

  def mapper(self, _, line):
    splits = line.rstrip("\n").split("|")

    if len(splits) == 2: # country data
      symbol = 'A' # make country sort before transaction data
      country2digit = splits[1]
      yield country2digit, [symbol, splits]
    else: # person data
      symbol = 'B'
      country2digit = splits[2]
      yield country2digit, [symbol, splits]

  def reducer(self, key, values):
    countries = [] # should come first, as they are sorted on artificia key 'A'
    for value in values:
      if value[0] == 'A':
        countries.append(value)
      if value[0] == 'B':
        for country in countries:
          yield key, country[1:] + value[1:]

if __name__ == '__main__':
  MRJoin.run()

# 5.3 Pairwise similarity  - PHASE 1 <a name="5.3"></a>

In this part of the assignment we will focus on developing methods for detecting synonyms, using the Google 5-grams dataset. To accomplish this you must script two main tasks using MRJob:


#### (1) Using the systems tests data sets, write mrjob code to build the stripes
#### (2) Write mrjob code to build an inverted index from the stripes
#### (3) Using two (symmetric) comparison methods of your choice (e.g., correlations, distances, similarities), pairwise compare all stripes (vectors), and output to a file.   

__==Design notes for (1)== __  
For this task you will be able to modify the pattern we used in HW 3.2 (feel free to use the solution as reference). To total the word counts across the n-grams, output the support from the mappers using the total order inversion pattern:

<*word,count>   

to ensure that the support arrives before the cooccurrences.   

In addition to ensuring the determination of the total word counts, the mapper must also output co-occurrence counts for the pairs of words inside of each n-gram. Treat these words as a basket, as we have in HW 3, but count all stripes or pairs in both orders, i.e., count both orderings: (word1,word2), and (word2,word1), to preserve
symmetry in our output for (2).

__==Design notes for (3)==__   
For this task you will have to determine a method of comparison.
Here are a few that you might consider:

 - Jaccard
 - Cosine similarity
 - Spearman correlation
 - Euclidean distance
 - Taxicab (Manhattan) distance
 - Shortest path graph distance (a graph, because our data is symmetric!)
 - Pearson correlation
 - Kendall correlation
 ...

However, be cautioned that some comparison methods are more difficult to parallelize than others, and do not perform more associations than is necessary, since your choice of association will be symmetric.

Please use the inverted index (discussed in live session #5) based pattern to compute the pairwise (term-by-term) similarity matrix. 

In [319]:
%%writefile buildStripes.py
#!~/anaconda2/bin/python
# -*- coding: utf-8 -*-

from __future__ import division
import re
import mrjob
import json
from mrjob.protocol import RawProtocol
from mrjob.job import MRJob
from mrjob.step import MRStep
import itertools
import collections as cl
import ast

class MRbuildStripes(MRJob):
  
    #START SUDENT CODE531_STRIPES
    
    #OUTPUT_PROTOCOL = JSONValueProtocol
    
    def steps(self):
        
        JOBCONF_STEP = {
            'mapreduce.job.output.key.comparator.class': 'org.apache.hadoop.mapred.lib.KeyFieldBasedComparator',
            'stream.map.output.field.separator':'\t',    
            'mapreduce.partition.keycomparator.options': '-k1,1',
            'mapreduce.job.reduces': '1'
        }
        
        return [
            MRStep(jobconf=JOBCONF_STEP,
                   mapper=self.mapper,
                   reducer_init=self.reducer_init,
                   reducer=self.reducer
                  )
        ]
    
    def mapper(self, _, line):
        
        #instantiate dict to hold stripes
        stripes_dict = {}
        
        #parse the co-occurrence count from the line
        co_occur_count = line.strip().split()[-3]

        #get pairs of words in the 5-gram and create a stripe for each unique word
        #Note the set here ensures that a word that appears more than once in a 5-gram won't be counted as co-occurring with itself
        #Note the permutations here will count both (word1,word2) and (word2,word1)
        for subset in itertools.permutations(sorted(set(line.strip().lower().split()[:-3])), 2):
            if subset[0] not in stripes_dict.keys():
                stripes_dict[subset[0]] = {}
                stripes_dict[subset[0]][subset[1]] = int(co_occur_count)
                
            elif subset[1] not in stripes_dict[subset[0]]:
                stripes_dict[subset[0]][subset[1]] = int(co_occur_count)
                
            else:
                stripes_dict[subset[0]][subset[1]] += int(co_occur_count)
        
        for key in stripes_dict:
            #yield key, json.dumps(stripes_dict[key])
            yield key, json.dumps(stripes_dict[key])

    def reducer_init(self):
        #instantiate placholder current key and counter
        self.placeholder_for_reducer = 0
        
    def reducer(self,key,stripe):
        
        #instantiate counter to hold combined stripes
        cur_counter = cl.Counter()
                
        for stripe_dict in stripe:
            cur_key = key
            
            #load the stripe into a dict
            stripe_dict_to_add = json.loads(stripe_dict)
            
            #update with each stripe
            cur_counter.update(stripe_dict_to_add)

        #output the key and the counter with the sum of all co-occurrences for that key
        yield key, cur_counter
        
        #The placeholder only increments when a new key is encountered
        #reducer seems to get chunks of key value pairs from the mapper with the same key all together
        self.placeholder_for_reducer += 1

    #END SUDENT CODE531_STRIPES

if __name__ == '__main__':
    MRbuildStripes.run()

Overwriting buildStripes.py


In [200]:
%%writefile invertedIndex.py
#!~/anaconda2/bin/python
# -*- coding: utf-8 -*-

from __future__ import division
import collections as cl
import re
import json
import math
import numpy as np
import itertools
import mrjob
from mrjob.protocol import RawProtocol
from mrjob.job import MRJob
from mrjob.step import MRStep
import sys

class MRinvertedIndex(MRJob):
    #START SUDENT CODE531_INV_INDEX
    
    #OUTPUT_PROTOCOL = JSONValueProtocol

    
    def steps(self):
        
        JOBCONF_STEP = {
            'mapreduce.job.output.key.comparator.class': 'org.apache.hadoop.mapred.lib.KeyFieldBasedComparator',
            'stream.map.output.field.separator':'\t',    
            'mapreduce.partition.keycomparator.options': '-k1,1',
            'mapreduce.job.reduces': '1'
        }
        
        return [
            MRStep(jobconf=JOBCONF_STEP,
                   mapper=self.mapper,
                   reducer=self.reducer
                  )
        ]
    
    def mapper(self, _, line):
        
        #Read the key and the stripe from the input
        key_stripe_list = line.translate(None,"\n").split("\t")
        
        key = json.loads(key_stripe_list[0])
        stripe = json.loads(key_stripe_list[1])
        
        #get length of the stripe
        stripe_len = len(stripe)
        
        #create dict that will hold doc title and the stripe length
        stripe_len_dict = {key:stripe_len}
        
        for term in stripe:
            yield term, stripe_len_dict
  
    def reducer(self, key, stripe):

        #instantiate a dict where we will combine stripe_len_dicts emitted from mapper
        stripe_len_dict_combined = {}
        
        #Read the incremental stripe length dit from the mapper output
        for stripe_len_dict in stripe:
            
            #update the dict above with each incremental stripe length dict
            stripe_len_dict_combined.update(stripe_len_dict)
                    
        yield key, stripe_len_dict_combined        
        
        #END SUDENT CODE531_INV_INDEX
        
if __name__ == '__main__':
    MRinvertedIndex.run() 

Overwriting invertedIndex.py


In [360]:
%%writefile similarity.py
#!~/anaconda2/bin/python
# -*- coding: utf-8 -*-

from __future__ import division
import collections
import re
import json
import math
import numpy as np
import itertools
import mrjob
from mrjob.protocol import RawProtocol
from mrjob.job import MRJob
from mrjob.step import MRStep

class MRsimilarity(MRJob):
  
    #START SUDENT CODE531_SIMILARITY
    #OUTPUT_PROTOCOL = JSONValueProtocol
    
    def steps(self):
        
        JOBCONF_STEP = {
            'mapreduce.job.output.key.comparator.class': 'org.apache.hadoop.mapred.lib.KeyFieldBasedComparator',
            'stream.map.output.field.separator':'\t',    
            'stream.num.map.output.key.fields': '2',
            'mapreduce.partition.keycomparator.options': '-k1,1',
            'mapreduce.job.reduces': '1'
        }
        
        return [
            MRStep(jobconf=JOBCONF_STEP,
                   mapper=self.mapper,
                   reducer=self.reducer
                  )
        ]
    
    def mapper(self, _, line):
        
        #Read the key and the stripe from the input
        key_index_list = line.translate(None,"\n").split("\t")
        
        key = json.loads(key_index_list[0])
        inv_index = json.loads(key_index_list[1])
        
        for subset in itertools.permutations(sorted(set(inv_index.keys())), 2):
            
            #get the lengths of each item that we want to output
            item_1_len = inv_index[subset[0]]
            item_2_len = inv_index[subset[1]]
            
            #create a two item dict to hold the item names and their lengths
            item_len_dict = {}
            item_len_dict[subset[0]] = item_1_len
            item_len_dict[subset[1]] = item_2_len
            
            #when output like this, can index with key[0] to get subset[0] in the reducer
            yield (subset[0], subset[1]), (1,item_len_dict)
  
    def reducer(self, key, stripe):

        #instantiate object to hold the number of times the terms / documents intersect
        items_intersect = 0
        cosine_sim_product = 0
        
        #the reducer gets output from the mapper in chunks grouped by key
        for n,item_len_dict in stripe:
            
            #get the length of item 1 and item 2
            item_1_len = item_len_dict[key[0]]
            item_2_len = item_len_dict[key[1]]
            
            #increment the intersect count
            items_intersect = items_intersect + n
            
            cosine_sim_product = cosine_sim_product + (1/np.sqrt(item_1_len)) * (1/np.sqrt(item_2_len))
            
        #calculate jacard similarity score
        jacard_sim_score = float(items_intersect)/ float(item_1_len + item_2_len - items_intersect)
        
        #Record cosine similarity score before moving on to the next chunk
        cosine_sim_score = cosine_sim_product
        
        yield key, (jacard_sim_score, cosine_sim_score)   
    
    #END SUDENT CODE531_SIMILARITY

if __name__ == '__main__':
    MRsimilarity.run()

Overwriting similarity.py


## HW5.3.1   Run Systems tests locally on small datasets (PHASE1) <a name="5.3.1"></a>  
[Back to Table of Contents](#TOC)

Complete 5.3 and systems test using the below test datasets. Phase 2 will focus on the entire Ngram dataset.

To help you through these tasks please verify that your code gives the results below (for stripes, inverted index, and pairwise similarities).

Test datasets:

* googlebooks-eng-all-5gram-20090715-0-filtered.txt [see below]
* atlas-boon-test [see below]
* stripe-docs-test [see below]


A large subset of the Google n-grams dataset

https://aws.amazon.com/datasets/google-books-ngrams/

which we have placed in a bucket/folder on Dropbox and on s3:

https://www.dropbox.com/sh/tmqpc4o0xswhkvz/AACUifrl6wrMrlK6a3X3lZ9Ea?dl=0 

s3://filtered-5grams/

In particular, this bucket contains (~200) files (10Meg each) in the format:

	(ngram) \t (count) \t (pages_count) \t (books_count)

The next cell shows the first 10 lines of the googlebooks-eng-all-5gram-20090715-0-filtered.txt file.


__DISCLAIMER__: Each record is already a 5-gram. In real life, we would calculate the stripes cooccurrence data from the raw text by windowing over the raw text and not from the 5-gram preprocessed data (as we are doing here).  Calculatating pairs on this 5-gram is a little corrupt as we will be double counting cooccurences. Having said that this exercise can still pull out some simialr terms. 

#### 1: unit/systems first-10-lines

In [231]:
%%writefile googlebooks-eng-all-5gram-20090715-0-filtered-first-10-lines.txt
A BILL FOR ESTABLISHING RELIGIOUS	59	59	54
A Biography of General George	92	90	74
A Case Study in Government	102	102	78
A Case Study of Female	447	447	327
A Case Study of Limited	55	55	43
A Child's Christmas in Wales	1099	1061	866
A Circumstantial Narrative of the	62	62	50
A City by the Sea	62	60	49
A Collection of Fairy Tales	123	117	80
A Collection of Forms of	116	103	82

Overwriting googlebooks-eng-all-5gram-20090715-0-filtered-first-10-lines.txt


#### 2: unit/systems atlas-boon

In [153]:
%%writefile atlas-boon-systems-test.txt
atlas boon	50	50	50
boon cava dipped	10	10	10
atlas dipped	15	15	15

Overwriting atlas-boon-systems-test.txt


#### 3: unit/systems stripe-docs-test

Three terms, A,B,C and their corresponding stripe-docs of co-occurring terms

- DocA {X:20, Y:30, Z:5}
- DocB {X:100, Y:20}
- DocC {M:5, N:20, Z:5}

### (1) build stripes for all the test data sets - run the commands and insure that your output matches the output below

In [320]:
###########################################################################
# Make Stripes from ngrams for systems test 1
###########################################################################

!hdfs dfs rm --recursive systems_test_stripes_1

#Local works
!python buildStripes.py -r local googlebooks-eng-all-5gram-20090715-0-filtered-first-10-lines.txt > systems_test_stripes_1   

#Hadoop also works
#!python buildStripes.py -r hadoop googlebooks-eng-all-5gram-20090715-0-filtered-first-10-lines.txt > systems_test_stripes_1   



rm: Unknown command
Did you mean -rm?  This command begins with a dash.
Using configs in /etc/mrjob.conf
Creating temp directory /tmp/buildStripes.root.20170609.193359.649160
Running step 1 of 1...
Streaming final output from /tmp/buildStripes.root.20170609.193359.649160/output...
Removing temp directory /tmp/buildStripes.root.20170609.193359.649160...


In [321]:
!cat systems_test_stripes_1

"a"	{"limited":55,"female":447,"general":92,"sea":62,"in":1201,"religious":59,"george":92,"biography":92,"city":62,"for":59,"tales":123,"child's":1099,"forms":116,"wales":1099,"christmas":1099,"government":102,"collection":239,"by":62,"case":604,"circumstantial":62,"fairy":123,"of":895,"study":604,"bill":59,"establishing":59,"narrative":62,"the":124}
"bill"	{"a":59,"religious":59,"for":59,"establishing":59}
"biography"	{"a":92,"of":92,"george":92,"general":92}
"by"	{"a":62,"city":62,"the":62,"sea":62}
"case"	{"a":604,"limited":55,"government":102,"of":502,"study":604,"female":447,"in":102}
"child's"	{"a":1099,"wales":1099,"christmas":1099,"in":1099}
"christmas"	{"a":1099,"wales":1099,"in":1099,"child's":1099}
"circumstantial"	{"a":62,"of":62,"the":62,"narrative":62}
"city"	{"a":62,"the":62,"by":62,"sea":62}
"collection"	{"a":239,"forms":116,"fairy":123,"tales":123,"of":239}
"establishing"	{"a":59,"bill":59,"religious":59,"for":59}
"fairy"	{"a":123,"of":123,"tales":123,"collection":123}

<pre>
"a"	{"limited": 55, "sea": 62, "general": 92, "female": 447, "in": 1201, "religious": 59, "george": 92, "biography": 92, "city": 62, "for": 59, "tales": 123, "child's": 1099, "forms": 116, "wales": 1099, "christmas": 1099, "government": 102, "collection": 239, "by": 62, "case": 604, "circumstantial": 62, "fairy": 123, "of": 1011, "study": 604, "bill": 59, "establishing": 59, "narrative": 62, "the": 124}
"bill"	{"a": 59, "religious": 59, "for": 59, "establishing": 59}
"biography"	{"a": 92, "of": 92, "george": 92, "general": 92}
"by"	{"a": 62, "city": 62, "the": 62, "sea": 62}
"case"	{"a": 604, "limited": 55, "government": 102, "of": 502, "study": 604, "female": 447, "in": 102}
"child's"	{"a": 1099, "wales": 1099, "christmas": 1099, "in": 1099}
"christmas"	{"a": 1099, "wales": 1099, "in": 1099, "child's": 1099}
"circumstantial"	{"a": 62, "of": 62, "the": 62, "narrative": 62}
"city"	{"a": 62, "the": 62, "by": 62, "sea": 62}
"collection"	{"a": 239, "of": 355, "fairy": 123, "tales": 123, "forms": 116}
"establishing"	{"a": 59, "bill": 59, "religious": 59, "for": 59}
"fairy"	{"a": 123, "of": 123, "tales": 123, "collection": 123}
"female"	{"a": 447, "case": 447, "study": 447, "of": 447}
"for"	{"a": 59, "bill": 59, "religious": 59, "establishing": 59}
"forms"	{"a": 116, "of": 232, "collection": 116}
"general"	{"a": 92, "of": 92, "george": 92, "biography": 92}
"george"	{"a": 92, "of": 92, "biography": 92, "general": 92}
"government"	{"a": 102, "case": 102, "study": 102, "in": 102}
"in"	{"a": 1201, "case": 102, "government": 102, "study": 102, "child's": 1099, "wales": 1099, "christmas": 1099}
"limited"	{"a": 55, "case": 55, "study": 55, "of": 55}
"narrative"	{"a": 62, "of": 62, "the": 62, "circumstantial": 62}
"of"	{"a": 1127, "case": 502, "circumstantial": 62, "george": 92, "limited": 55, "tales": 123, "collection": 471, "general": 92, "forms": 348, "female": 447, "narrative": 62, "study": 502, "fairy": 123, "the": 62, "biography": 92}
"religious"	{"a": 59, "bill": 59, "for": 59, "establishing": 59}
"sea"	{"a": 62, "city": 62, "the": 62, "by": 62}
"study"	{"a": 604, "case": 604, "limited": 55, "government": 102, "of": 502, "female": 447, "in": 102}
"tales"	{"a": 123, "of": 123, "fairy": 123, "collection": 123}
"the"	{"a": 124, "city": 62, "circumstantial": 62, "of": 62, "sea": 62, "narrative": 62, "by": 62}
"wales"	{"a": 1099, "in": 1099, "christmas": 1099, "child's": 1099}

</pre>

In [324]:
###########################################################################
# Make Stripes from ngrams for systems test 2
###########################################################################

!hdfs dfs rm --recursive systems_test_stripes_2

#Local mode works
#!python buildStripes.py -r local atlas-boon-systems-test.txt > systems_test_stripes_2

#Hadoop mode
!python buildStripes.py -r hadoop atlas-boon-systems-test.txt > systems_test_stripes_2

rm: Unknown command
Did you mean -rm?  This command begins with a dash.
Using configs in /etc/mrjob.conf
Looking for hadoop binary in $PATH...
Found hadoop binary: /usr/bin/hadoop
Using Hadoop version 2.6.0
Looking for Hadoop streaming jar in /home/hadoop/contrib...
Looking for Hadoop streaming jar in /usr/lib/hadoop-mapreduce...
Found Hadoop streaming jar: /usr/lib/hadoop-mapreduce/hadoop-streaming.jar
Creating temp directory /tmp/buildStripes.root.20170609.193705.029966
Copying local files to hdfs:///user/root/tmp/mrjob/buildStripes.root.20170609.193705.029966/files/...
Running step 1 of 1...
  packageJobJar: [] [/usr/jars/hadoop-streaming-2.6.0-cdh5.7.0.jar] /tmp/streamjob6846033796536965021.jar tmpDir=null
  Connecting to ResourceManager at /0.0.0.0:8032
  Connecting to ResourceManager at /0.0.0.0:8032
  Total input paths to process : 1
  number of splits:2
  Submitting tokens for job: job_1497021168065_0003
  Submitted application application_1497021168065_0003
  The url to track 

In [325]:
!cat systems_test_stripes_2

"atlas"	{"dipped":15,"boon":50}
"boon"	{"atlas":50,"dipped":10,"cava":10}
"cava"	{"dipped":10,"boon":10}
"dipped"	{"atlas":15,"boon":10,"cava":10}


<pre>
"atlas"   {"dipped": 15, "boon": 50}   
"boon"    {"atlas": 50, "dipped": 10, "cava": 10}   
"cava"    {"dipped": 10, "boon": 10} 
"dipped"  {"atlas": 15, "boon": 10, "cava": 10}
</pre>

In [195]:
########################################################################
# Stripes for systems test 3 (given, no need to build stripes)
########################################################################

with open("systems_test_stripes_3", "w") as f:
    f.writelines([
        '"DocA"\t{"X":20, "Y":30, "Z":5}\n',
        '"DocB"\t{"X":100, "Y":20}\n',  
        '"DocC"\t{"M":5, "N":20, "Z":5, "Y":1}\n'
    ])
!cat systems_test_stripes_3   

"DocA"	{"X":20, "Y":30, "Z":5}
"DocB"	{"X":100, "Y":20}
"DocC"	{"M":5, "N":20, "Z":5, "Y":1}


### (2) Build Inverted Index - run the commands and insure that your output matches the output below

In [206]:
#Local works
#!python invertedIndex.py -r local --file=systems_test_stripes_1  systems_test_stripes_1 > systems_test_index_1
!python invertedIndex.py -r local systems_test_stripes_1 > systems_test_index_1


#Hadoop mode also works
#!python invertedIndex.py -r hadoop systems_test_stripes_1 > systems_test_index_1


Using configs in /etc/mrjob.conf
Creating temp directory /tmp/invertedIndex.root.20170610.191444.468690
Running step 1 of 1...
Streaming final output from /tmp/invertedIndex.root.20170610.191444.468690/output...
Removing temp directory /tmp/invertedIndex.root.20170610.191444.468690...


In [207]:
#Local works
!python invertedIndex.py -r local systems_test_stripes_2 > systems_test_index_2

#Hadoop mode also works
#!python invertedIndex.py -r hadoop systems_test_stripes_2 > systems_test_index_2


Using configs in /etc/mrjob.conf
Creating temp directory /tmp/invertedIndex.root.20170610.191446.131567
Running step 1 of 1...
Streaming final output from /tmp/invertedIndex.root.20170610.191446.131567/output...
Removing temp directory /tmp/invertedIndex.root.20170610.191446.131567...


In [208]:
#Local works
!python invertedIndex.py -r local systems_test_stripes_3 > systems_test_index_3

#Hadoop mode also works
#!python invertedIndex.py -r hadoop systems_test_stripes_3 > systems_test_index_3


Using configs in /etc/mrjob.conf
Creating temp directory /tmp/invertedIndex.root.20170610.191503.808725
Running step 1 of 1...
Streaming final output from /tmp/invertedIndex.root.20170610.191503.808725/output...
Removing temp directory /tmp/invertedIndex.root.20170610.191503.808725...


In [240]:
##########################################################
# Pretty print systems tests for generating Inverted Index
##########################################################

import json

for i in range(1,4):
    print "—"*100
    print "Systems test ",i," - Inverted Index"
    print "—"*100  
    with open("systems_test_index_"+str(i),"r") as f:
        lines = f.readlines()
        for line in lines:
            line = line.strip()
            word,stripe = line.split("\t")
            stripe = json.loads(stripe)
            #stripe.extend([["",""] for _ in xrange(3 - len(stripe))])
            if len(stripe) > 2:
                print "{0:>16} |{1:>16} |{2:>16} |{3:>16}".format(
                (word), stripe.keys()[0]+" "+str(stripe[stripe.keys()[0]]), stripe.keys()[1]+" "+str(stripe[stripe.keys()[1]]), stripe.keys()[2]+" "+str(stripe[stripe.keys()[2]]))    
            elif len(stripe) == 2:
                print "{0:>16} |{1:>16} |{2:>16} |{3:>16}".format(
                (word), stripe.keys()[0]+" "+str(stripe[stripe.keys()[0]]), stripe.keys()[1]+" "+str(stripe[stripe.keys()[1]]), " ")
            else:
                print "{0:>16} |{1:>16} |{2:>16} |{3:>16}".format(
                (word), stripe.keys()[0]+" "+str(stripe[stripe.keys()[0]])," ", " ")


————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  1  - Inverted Index
————————————————————————————————————————————————————————————————————————————————————————————————————
             "a" |       limited 4 |           sea 4 |       general 4
          "bill" |            a 27 |     religious 4 |           for 4
     "biography" |            a 27 |           of 15 |        george 4
            "by" |            a 27 |          city 4 |           the 7
          "case" |            a 27 |       limited 4 |    government 4
       "child's" |            a 27 |         wales 4 |     christmas 4
     "christmas" |            a 27 |         wales 4 |       child's 4
"circumstantial" |            a 27 |           of 15 |           the 7
          "city" |            a 27 |           the 7 |            by 4
    "collection" |            a 27 |         forms 3 |         fairy 4
  "establishing" |            a 27 |          bill 4 |

### Inverted Index

In [ ]:
————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  1  - Inverted Index
————————————————————————————————————————————————————————————————————————————————————————————————————
             "a" |          bill 4 |     biography 4 |            by 4
          "bill" |            a 27 |  establishing 4 |           for 4
     "biography" |            a 27 |       general 4 |        george 4
            "by" |            a 27 |          city 4 |           sea 4
          "case" |            a 27 |        female 4 |    government 4
       "child's" |            a 27 |     christmas 4 |            in 7
     "christmas" |            a 27 |       child's 4 |            in 7
"circumstantial" |            a 27 |     narrative 4 |           of 15
          "city" |            a 27 |            by 4 |           sea 4
    "collection" |            a 27 |         fairy 4 |         forms 3
  "establishing" |            a 27 |          bill 4 |           for 4
         "fairy" |            a 27 |    collection 5 |           of 15
        "female" |            a 27 |          case 7 |           of 15
           "for" |            a 27 |          bill 4 |  establishing 4
         "forms" |            a 27 |    collection 5 |           of 15
       "general" |            a 27 |     biography 4 |        george 4
        "george" |            a 27 |     biography 4 |       general 4
    "government" |            a 27 |          case 7 |            in 7
            "in" |            a 27 |          case 7 |       child's 4
       "limited" |            a 27 |          case 7 |           of 15
     "narrative" |            a 27 |circumstantial 4 |           of 15
            "of" |            a 27 |     biography 4 |          case 7
     "religious" |            a 27 |          bill 4 |  establishing 4
           "sea" |            a 27 |            by 4 |          city 4
         "study" |            a 27 |          case 7 |        female 4
         "tales" |            a 27 |    collection 5 |         fairy 4
           "the" |            a 27 |            by 4 |circumstantial 4
         "wales" |            a 27 |       child's 4 |     christmas 4
————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  2  - Inverted Index
————————————————————————————————————————————————————————————————————————————————————————————————————
         "atlas" |          boon 3 |        dipped 3 |                
          "boon" |         atlas 2 |          cava 2 |        dipped 3
          "cava" |          boon 3 |        dipped 3 |                
        "dipped" |         atlas 2 |          boon 3 |          cava 2
————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  3  - Inverted Index
————————————————————————————————————————————————————————————————————————————————————————————————————
             "M" |          DocC 4 |                 |                
             "N" |          DocC 4 |                 |                
             "X" |          DocA 3 |          DocB 2 |                
             "Y" |          DocA 3 |          DocB 2 |          DocC 4
             "Z" |          DocA 3 |          DocC 4 |                


### (3) Calculate similarities - run the commands and insure that your output matches the output below

#### NOTE: you must run in hadoop mode to generate sorted similarities

In [393]:
#Hadoop mode also works
!python similarity.py -r hadoop systems_test_index_1 > systems_test_similarities_1

Using configs in /etc/mrjob.conf
Looking for hadoop binary in $PATH...
Found hadoop binary: /usr/bin/hadoop
Using Hadoop version 2.6.0
Looking for Hadoop streaming jar in /home/hadoop/contrib...
Looking for Hadoop streaming jar in /usr/lib/hadoop-mapreduce...
Found Hadoop streaming jar: /usr/lib/hadoop-mapreduce/hadoop-streaming.jar
Creating temp directory /tmp/similarity.root.20170612.022355.751309
Copying local files to hdfs:///user/root/tmp/mrjob/similarity.root.20170612.022355.751309/files/...
Running step 1 of 1...
  packageJobJar: [] [/usr/jars/hadoop-streaming-2.6.0-cdh5.7.0.jar] /tmp/streamjob4205294873810135897.jar tmpDir=null
  Connecting to ResourceManager at /0.0.0.0:8032
  Connecting to ResourceManager at /0.0.0.0:8032
  Total input paths to process : 1
  number of splits:2
  Submitting tokens for job: job_1497107873528_0029
  Submitted application application_1497107873528_0029
  The url to track the job: http://quickstart.cloudera:8088/proxy/application_1497107873528_002

In [363]:
#Local mode
!python similarity.py -r hadoop systems_test_index_2 > systems_test_similarities_2

Using configs in /etc/mrjob.conf
Looking for hadoop binary in $PATH...
Found hadoop binary: /usr/bin/hadoop
Using Hadoop version 2.6.0
Looking for Hadoop streaming jar in /home/hadoop/contrib...
Looking for Hadoop streaming jar in /usr/lib/hadoop-mapreduce...
Found Hadoop streaming jar: /usr/lib/hadoop-mapreduce/hadoop-streaming.jar
Creating temp directory /tmp/similarity.root.20170611.023918.305852
Copying local files to hdfs:///user/root/tmp/mrjob/similarity.root.20170611.023918.305852/files/...
Running step 1 of 1...
  packageJobJar: [] [/usr/jars/hadoop-streaming-2.6.0-cdh5.7.0.jar] /tmp/streamjob8878015087281128230.jar tmpDir=null
  Connecting to ResourceManager at /0.0.0.0:8032
  Connecting to ResourceManager at /0.0.0.0:8032
  Total input paths to process : 1
  number of splits:2
  Submitting tokens for job: job_1497107873528_0027
  Submitted application application_1497107873528_0027
  The url to track the job: http://quickstart.cloudera:8088/proxy/application_1497107873528_002

In [364]:
!python similarity.py -r hadoop systems_test_index_3 > systems_test_similarities_3

Using configs in /etc/mrjob.conf
Looking for hadoop binary in $PATH...
Found hadoop binary: /usr/bin/hadoop
Using Hadoop version 2.6.0
Looking for Hadoop streaming jar in /home/hadoop/contrib...
Looking for Hadoop streaming jar in /usr/lib/hadoop-mapreduce...
Found Hadoop streaming jar: /usr/lib/hadoop-mapreduce/hadoop-streaming.jar
Creating temp directory /tmp/similarity.root.20170611.024046.301768
Copying local files to hdfs:///user/root/tmp/mrjob/similarity.root.20170611.024046.301768/files/...
Running step 1 of 1...
  packageJobJar: [] [/usr/jars/hadoop-streaming-2.6.0-cdh5.7.0.jar] /tmp/streamjob4692597056649005003.jar tmpDir=null
  Connecting to ResourceManager at /0.0.0.0:8032
  Connecting to ResourceManager at /0.0.0.0:8032
  Total input paths to process : 1
  number of splits:2
  Submitting tokens for job: job_1497107873528_0028
  Submitted application application_1497107873528_0028
  The url to track the job: http://quickstart.cloudera:8088/proxy/application_1497107873528_002

In [72]:
############################################
# Pretty print systems tests
############################################

import json
for i in range(1,4):
  print '—'*110
  print "Systems test ",i," - Similarity measures"
  print '—'*110
  print "{0:>15} |{1:>15} |{2:>15} |{3:>15} |{4:>15} |{5:>15}".format(
          "average", "pair", "cosine", "jaccard", "overlap", "dice")
  print '-'*110

  with open("systems_test_similarities_"+str(i),"r") as f:
      lines = f.readlines()
      for line in lines:
          line = line.strip()
          avg,stripe = line.split("\t")
          stripe = json.loads(stripe)

          print "{0:>15f} |{1:>15} |{2:>15f} |{3:>15f} |{4:>15f} |{5:>15f}".format(
              float(avg), stripe[0], float(stripe[1]), float(stripe[2]), float(stripe[3]), float(stripe[4]))
 

In [365]:
!cat systems_test_similarities_2

["atlas","boon"]	[0.25,0.4082482905]
["atlas","cava"]	[1.0,1.0]
["atlas","dipped"]	[0.25,0.4082482905]
["boon","atlas"]	[0.25,0.4082482905]
["boon","cava"]	[0.25,0.4082482905]
["boon","dipped"]	[0.5,0.6666666667]
["cava","atlas"]	[1.0,1.0]
["cava","boon"]	[0.25,0.4082482905]
["cava","dipped"]	[0.25,0.4082482905]
["dipped","atlas"]	[0.25,0.4082482905]
["dipped","boon"]	[0.5,0.6666666667]
["dipped","cava"]	[0.25,0.4082482905]


In [392]:
#Pretty print the similarities

import ast

for i in range(1,4):

    #Print titles
    print '—'*110
    print "Systems test ",i," - Similarity measures"
    print '—'*110
    print "{0:>25} |{1:>15} |{2:>15}".format(
          "pair", "jaccard", "cosine")
    print '-'*110

    #read file and keep only first 15 lines
    f = open('systems_test_similarities_'+str(i)).readlines()
    first_15 = f[:15]

    #print the first 15 lines of each output
    for line in first_15:
        terms_sims_list = line.split('\t')
        sims = ast.literal_eval(terms_sims_list[1])
        #print the calculated similarities
        print "{0:>25s} |{1:>15f} |{2:>15f}".format(
                  terms_sims_list[0], sims[0], sims[1])

 

——————————————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  1  - Similarity measures
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
                     pair |        jaccard |         cosine
--------------------------------------------------------------------------------------------------------------
             ["a","bill"] |       0.107143 |       0.288675
        ["a","biography"] |       0.107143 |       0.288675
               ["a","by"] |       0.107143 |       0.288675
             ["a","case"] |       0.214286 |       0.436436
          ["a","child's"] |       0.107143 |       0.288675
        ["a","christmas"] |       0.107143 |       0.288675
   ["a","circumstantial"] |       0.107143 |       0.288675
             ["a","city"] |       0.107143 |       0.288675
       ["a","collection"] |       0.142857 |       0.344265
     ["a","establishing"] | 

### Pairwise Similairity 

In [ ]:
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  1  - Similarity measures
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
   average |                pair |         cosine |        jaccard |        overlap |           dice
--------------------------------------------------------------------------------------------------------------
  1.000000 |    female - limited |       1.000000 |       1.000000 |       1.000000 |       1.000000
  0.868292 |       fairy - forms |       0.866025 |       0.750000 |       1.000000 |       0.857143
  0.868292 |       forms - tales |       0.866025 |       0.750000 |       1.000000 |       0.857143
  0.830357 |        case - study |       0.857143 |       0.750000 |       0.857143 |       0.857143
  0.712500 | bill - establishing |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |   christmas - wales |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |circumstantial - narrative |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |            by - sea |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |           by - city |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |     child's - wales |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |  biography - george |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 | child's - christmas |       0.750000 |       0.600000 |       0.750000 |       0.750000
  ...
  
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  2  - Similarity measures
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
   average |                pair |         cosine |        jaccard |        overlap |           dice
--------------------------------------------------------------------------------------------------------------
  1.000000 |        atlas - cava |       1.000000 |       1.000000 |       1.000000 |       1.000000
  0.625000 |       boon - dipped |       0.666667 |       0.500000 |       0.666667 |       0.666667
  0.389562 |       cava - dipped |       0.408248 |       0.250000 |       0.500000 |       0.400000
  0.389562 |         boon - cava |       0.408248 |       0.250000 |       0.500000 |       0.400000
  0.389562 |      atlas - dipped |       0.408248 |       0.250000 |       0.500000 |       0.400000
  0.389562 |        atlas - boon |       0.408248 |       0.250000 |       0.500000 |       0.400000
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  3  - Similarity measures
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
   average |                pair |         cosine |        jaccard |        overlap |           dice
--------------------------------------------------------------------------------------------------------------
  0.820791 |         DocA - DocB |       0.816497 |       0.666667 |       1.000000 |       0.800000
  0.553861 |         DocA - DocC |       0.577350 |       0.400000 |       0.666667 |       0.571429
  0.346722 |         DocB - DocC |       0.353553 |       0.200000 |       0.500000 |       0.333333


# === END OF PHASE 1 ===